crates/uv-pep508/src/marker/tree.rs

```diff
@@ -1153,6 +1153,49 @@ impl MarkerTree {
         MarkerTree(INTERNER.lock().only_extras(self.0))
     }
 
+    /// Calls the provided function on every `extra` in this tree.
+    ///
+    /// The operator provided to the function is guaranteed to be
+    /// `MarkerOperator::Equal` or `MarkerOperator::NotEqual`.
+    pub fn visit_extras(self, mut f: impl FnMut(MarkerOperator, &ExtraName)) {
+        fn imp(tree: MarkerTree, f: &mut impl FnMut(MarkerOperator, &ExtraName)) {
+            match tree.kind() {
+                MarkerTreeKind::True | MarkerTreeKind::False => {}
+                MarkerTreeKind::Version(kind) => {
+                    for (tree, _) in simplify::collect_edges(kind.edges()) {
+                        imp(tree, f);
+                    }
+                }
+                MarkerTreeKind::String(kind) => {
+                    for (tree, _) in simplify::collect_edges(kind.children()) {
+                        imp(tree, f);
+                    }
+                }
+                MarkerTreeKind::In(kind) => {
+                    for (_, tree) in kind.children() {
+                        imp(tree, f);
+                    }
+                }
+                MarkerTreeKind::Contains(kind) => {
+                    for (_, tree) in kind.children() {
+                        imp(tree, f);
+                    }
+                }
+                MarkerTreeKind::Extra(kind) => {
+                    if kind.low.is_false() {
+                        f(MarkerOperator::Equal, kind.name().extra());
+                    } else {
+                        f(MarkerOperator::NotEqual, kind.name().extra());
+                    }
+                    for (_, tree) in kind.children() {
+                        imp(tree, f);
+                    }
+                }
+            }
+        }
+        imp(self, &mut f);
+    }
```

Comment:
@ibraheemdev Is this the best way to do this? And is it even right? I spent an embarrassing amount of time on just this part. (I originally wrote it by walking the internal representation, but that ended up being wrong in a way that I didn't understand.)

Basically, the essential challenge I faced here was figuring out whether I was looking at `extra == '...'` or `extra != '...'`.

Reply:
I think the way you have it written now will visit `extra == 'x'` if there is no possible solution to the tree when `extra != 'x'`. And it will visit `extra != 'x'` when the tree has a potential solution when `extra != 'x'`. I'm not sure whether you meant for those checks to be mutually exclusive. Should `(extra == 'x' and a) or (extra != 'x' and b)` visit both `extra == 'x'` and `extra != 'x'`?

Reply:
Yeah this is tricky. I'm not quite sure how to solve this. My _suspicion_ is that those types of expressions don't actually appear in this context. Because it would imply that there's a fork that sometimes _includes_ `x` and sometimes _excludes_ `x`. I believe that by construction that can't happen.

That's a somewhat precarious invariant to rely on, but I'm not quite sure how to fix it.